### PR TITLE
boards/nucleo-l4r5zi: initial basic support

### LINKED ISSUE #6629: Add support to (all) ST Nucleo boards
There's an ongoing work which goal is to add support for (all) [ST Nucleo boards](http://www.st.com/en/evaluation-tools/stm32-mcu-nucleo.html) to RIOT.

The goal if this issue is to keep track of this work. For all boards, it lists some information, the actual support status (available upstream, available PR, not supported) and give the link to the documentation in the Wiki (if any).

| # | Board | Board Family | MCU Family | Status | Documentation | 
| -- | -------- | ------------------ | ------------------ | -------- | --------------------- |
| 1 | [Nucleo-L011K4](http://www.st.com/en/evaluation-tools/nucleo-l011k4.html) |  Nucleo 32 |   M0+ |  Not supported | No doc | 
| 2 | [Nucleo-F031K6](http://www.st.com/en/evaluation-tools/nucleo-f031k6.html)  |  Nucleo 32 |  M0 |  PR #6345 | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo32-F031) | 
| 3 | [Nucleo-F042K6](http://www.st.com/en/evaluation-tools/nucleo-f042k6.html)  |  Nucleo 32 |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo32-F042) | 
| 4 | [Nucleo-L031K6](http://www.st.com/en/evaluation-tools/nucleo-l031k6.html) |  Nucleo 32 |  M0+ |  Not supported | No doc | 
| 5 | [Nucleo-F303K8](http://www.st.com/en/evaluation-tools/nucleo-f303k8.html) |  Nucleo 32 |  M4 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo32-F303) | 
| 6 | [Nucleo-L432KC](http://www.st.com/en/evaluation-tools/nucleo-l432kc.html) |  Nucleo 32 |  M4 |  Not supported | No doc | 
| 7 | [Nucleo-F030R8](http://www.st.com/en/evaluation-tools/nucleo-f030r8.html) |  Nucleo 64  |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F030) | 
| 8 | [Nucleo-F302R8](http://www.st.com/en/evaluation-tools/nucleo-f302r8.html) |  Nucleo 64 |  M4 | PR #6615 | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F302) | 
| 9 | [Nucleo-F334R8](http://www.st.com/en/evaluation-tools/nucleo-f334r8.html)  |  Nucleo 64 |  M4 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F334)  | 
| 10 | [Nucleo-L053R8](http://www.st.com/en/evaluation-tools/nucleo-l053r8.html)  |  Nucleo 64 |  M0+ |  PR #6628  |  No Doc | 
| 11 | [Nucleo-F070RB](http://www.st.com/en/evaluation-tools/nucleo-f070rb.html)   |  Nucleo 64 |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F070)  | 
| 12 | [Nucleo-F072RB](http://www.st.com/en/evaluation-tools/nucleo-f072rb.html)  |  Nucleo 64 |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F072)  | 
| 13 | [Nucleo-F103RB](http://www.st.com/en/evaluation-tools/nucleo-f103rb.html)  |  Nucleo 64 |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F103)  | 
| 14 | [Nucleo-F410RB](http://www.st.com/en/evaluation-tools/nucleo-f410rb.html)  |  Nucleo 64 |  M4 |  PR #6025  | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F410)  | 
| 15 | [Nucleo-L073RZ](http://www.st.com/en/evaluation-tools/nucleo-l073rz.html)  |  Nucleo 64 |  M0+ |  PR #6433  | No doc  | 
| 16 | [Nucleo-F091RC](http://www.st.com/en/evaluation-tools/nucleo-f091rc.html)  |  Nucleo 64 |  M0 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F091)  | 
| 17 | [Nucleo-F303RE](http://www.st.com/en/evaluation-tools/nucleo-f303re.html)  |  Nucleo 64 |  M4 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F303)  | 
| 18 | [Nucleo-L152RE](http://www.st.com/en/evaluation-tools/nucleo-l152re.html)  |  Nucleo 64 |  M3 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-L1)  | 
| 19 | [Nucleo-F401RE](http://www.st.com/en/evaluation-tools/nucleo-f401re.html)  |  Nucleo 64 |  M4 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F401)  | 
| 20 | [Nucleo-F446RE](http://www.st.com/en/evaluation-tools/nucleo-f446re.html)  |  Nucleo 64 |  M4 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F446)  | 
| 21 | [Nucleo-F411RE](http://www.st.com/en/evaluation-tools/nucleo-f411re.html)  |  Nucleo 64 |  M4 |  PR #5709  | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo-F411)  | 
| 22 | [Nucleo-L476RG](http://www.st.com/en/evaluation-tools/nucleo-l476rg.html)  |  Nucleo 64 |  M4 |  Not supported | No doc  | 
| 23 | [Nucleo-F303ZE](http://www.st.com/en/evaluation-tools/nucleo-f303ze.html)  |  Nucleo 144 |  M4 |  PR #6625  | No doc  | 
| 24 | [Nucleo-F446ZE](http://www.st.com/en/evaluation-tools/nucleo-f446ze.html)  |  Nucleo 144 |  M4 |  PR #6553  | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo144-F446)  | 
| 25 | [Nucleo-F722ZE](http://www.st.com/en/evaluation-tools/nucleo-f722ze.html)  |  Nucleo 144 |  M7 |  Not supported | No doc  | 
| 26 | [Nucleo-F207ZG](http://www.st.com/en/evaluation-tools/nucleo-f207zg.html)  |  Nucleo 144 |  M3 |  Supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo144-F207)  | 
| 27 | [Nucleo-F412ZG](http://www.st.com/en/evaluation-tools/nucleo-f412zg.html)  |  Nucleo 144 |  M4 |  Not supported | No doc  | 
| 28 | [Nucleo-F746ZG](http://www.st.com/en/evaluation-tools/nucleo-f746zg.html)  |  Nucleo 144 |  M7 |  Not supported | No doc  | 
| 29 | [Nucleo-F413ZH](http://www.st.com/en/evaluation-tools/nucleo-f413zh.html)  |  Nucleo 144 |  M4 |  PR #6565  | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo144-F413)  | 
| 30 | [Nucleo-F429ZI](http://www.st.com/en/evaluation-tools/nucleo-f429zi.html)  |  Nucleo 144 |  M4 |  Not supported | [Doc available](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Nucleo144-F429)  | 
| 31 | [Nucleo-F767ZI](http://www.st.com/en/evaluation-tools/nucleo-f767zi.html)  |  Nucleo 144 |  M7 |  Not supported | No Doc  |

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds a basic support to the [STM32 Nucleo-L4R5ZI](http://www.st.com/en/evaluation-tools/nucleo-l4r5zi.html) board:
- with one timer configured
- only one UART on the Arduino pinout is configured because of missing support for LPUART (something which is on my todo list)
- one SPI (not tested)

The auto memory length makefile was also updated for this family of L4 CPUs.

### Issues/PRs references

Related to #6629

### CODE REVIEW

**Comment by aabadie** [comment]
Forgot one important point, openocd on master fails to flash this board. It works with this patch: http://openocd.zylin.com/#/c/4491/

**Comment by aabadie** [comment]
#9158 is now merged, I'll try to adapt this PR to use it. The support for this CPU was also merged last week in openocd, see http://openocd.zylin.com/#/c/4777.

Rebase is also needed...

**Comment by fjmolinas** [comment]
@aabadie could you lend me this board to test?

File: cpu/stm32l4/include/cpu_conf.h [hunk 1]
```diff
@@ -52,6 +54,8 @@ extern "C" {
 #define CPU_IRQ_NUMOF                   (83U)
 #elif defined(CPU_MODEL_STM32L496ZG)
 #define CPU_IRQ_NUMOF                   (91U)
+#elif defined(CPU_MODEL_STM32L4R5ZI)
+#define CPU_IRQ_NUMOF                   (95U)
```
[the remark this answers is not in the document]
Comment (aabadie) [reply cpu/stm32l4/include/cpu_conf.h]: In the CMSIS, the last interrupt is index 94 (DMAMUX1_OVR_IRQn)

[the remark this answers is not in the document]
Comment (fjmolinas) [reply cpu/stm32l4/include/cpu_conf.h]: Yeah I realized that later, though I had removed this comment. Sorry

File: cpu/stm32l4/vectors.c [hunk 2]
```diff
@@ -210,12 +211,16 @@ ISR_VECTOR(1) const isr_t vector_cpu[CPU_IRQ_NUMOF] = {
...
     [SAI2_IRQn               ] = isr_sai2,                 /* [75] Serial Audio Interface 2 global interrupt */
 #endif
-#if defined(CPU_MODEL_STM32L432KC) || defined(CPU_MODEL_STM32L476RG) || \
-    defined(CPU_MODEL_STM32L476VG) || defined(CPU_MODEL_STM32L475VG) || \
-    defined(CPU_MODEL_STM32L496ZG)
+#if defined(CPU_MODEL_STM32L432KC) || defined(CPU_MODEL_STM32L433RC) || \
+    defined(CPU_MODEL_STM32L476RG) || defined(CPU_MODEL_STM32L476VG) || \
+    defined(CPU_MODEL_STM32L475VG) || defined(CPU_MODEL_STM32L496ZG)
     [TIM7_IRQn               ] = isr_tim7,                 /* [55] TIM7 global interrupt */
+    [QUADSPI_IRQn            ] = isr_quadspi,              /* [71] Quad SPI global interrupt */
```
[the remark this answers is not in the document]
Comment (aabadie) [reply cpu/stm32l4/vectors.c]: The vectors.c file is a bit a mess but I checked and it's fine.

[the remark this answers is not in the document]
Comment (fjmolinas) [reply cpu/stm32l4/vectors.c]: By duplicated I ment that there are two groups using the same but guarded differently.

File: cpu/stm32l4/vectors.c [hunk 3]
```diff
@@ -210,12 +211,16 @@ ISR_VECTOR(1) const isr_t vector_cpu[CPU_IRQ_NUMOF] = {
     [OTG_FS_IRQn             ] = isr_otg_fs,               /* [67] USB OTG FS global Interrupt */
     [SAI2_IRQn               ] = isr_sai2,                 /* [75] Serial Audio Interface 2 global interrupt */
 #endif
-#if defined(CPU_MODEL_STM32L432KC) || defined(CPU_MODEL_STM32L476RG) || \
-    defined(CPU_MODEL_STM32L476VG) || defined(CPU_MODEL_STM32L475VG) || \
-    defined(CPU_MODEL_STM32L496ZG)
+#if defined(CPU_MODEL_STM32L432KC) || defined(CPU_MODEL_STM32L433RC) || \
```
[the remark this answers is not in the document]
Comment (aabadie) [reply cpu/stm32l4/vectors.c]: In fact, it's part of e4395146f1d26e3090656cc0a20a68a2ea4f33f2 and it's fixing a missing irq function.

**Comment by aabadie** [comment]
> This PR also needs documentation.

Done. I think I also addressed your other comments.

File: boards/nucleo-l4r5zi/doc.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+## Overview
+
+The Nucleo-L4R5ZI is a board from ST's Nucleo family supporting a ARM Cortex-M4
+STM32L4R5ZI microcontroller with 640Kb of RAM and 2Mb of ROM Flash.
+
+## Flashing the device
+
+The ST Nucleo-L4R5ZI board includes an on-board ST-LINK programmer and can be
+flashed using OpenOCD.
+Note that the latest release of OpenOCD doesn't contain support for this board,
```
Comment (fjmolinas) [inline boards/nucleo-l4r5zi/doc.txt]: I would change this line for something like this:

```
@note The latest release of OpenOCD doesn't contain support for this board,
so a recent development version (including  http://openocd.zylin.com/#/c/4777)
must be built from source to be able to flash this board.
```

**Comment by aabadie** [comment]
> Final remarks regarding flashing requirements, squash directly afterwards.

Done (and last change before august ;) )

**Comment by fjmolinas** [comment]
> Done (and last change before august ;) )

Enjoy your vacations!

**Comment by fjmolinas** [comment]
Done (and last change before august ;) )

**Pull request merged**